### PR TITLE
Add tags multi-select to component form

### PR DESCRIPTION
**Pull request by chiaberry**
## Associated issues

https://github.com/cityofaustin/atd-data-tech/issues/12223

## Testing
**URL to test:** 
https://deploy-preview-1049--atd-moped-main.netlify.app/moped/projects

edited: Yall i just realized that you have to test this locally, because there are updates to the db

**Steps to test:**

Edit a component's attributes:

Screenshot 2023-05-15 at 4 26 44 PM

Add tags to the component. 
Save the component, come back and add more tags/ delete existing tags. 

---
#### Ship list
- [ ] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

### CODE REVIEW

File: moped-editor/src/views/projects/projectView/ProjectComponents/utils/form.js [hunk 1]
```diff
@@ -101,14 +101,35 @@ export const useSubphaseOptions = (subphases) =>
...
+ * @param {Object} data Data returned with moped_component_tags records
+ * @returns {Array} The options with value, label, and full data object to produce the component tags options
+ */
+export const useComponentTagsOptions = (data) =>
+  useMemo(() => {
+    if (!data) return [];
+
+    const options = data.moped_component_tags.map((tag) => ({
+      value: tag.id,
+      label: `${tag.type.charAt(0).toUpperCase() + tag.type.slice(1)} - ${
```
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectComponents/utils/form.js:115]: i like what you did here—but i also think we should fix this at the DB level. would you nix this the case handling and i'll fix the "type" values? i'm going to open a migration for that momentarily now that i have new tag values from NW.

File: moped-editor/src/views/projects/projectView/ProjectComponents/EditAttributesModal.js [hunk 2]
```diff
@@ -151,15 +159,20 @@ const EditAttributesModal = ({
     editDispatch({ type: "cancel_attributes_edit" });
   };
 
+  console.log(componentToEdit)
```
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectComponents/EditAttributesModal.js:162]: ❌

File: moped-editor/src/views/projects/projectView/ProjectComponents/EditAttributesModal.js [hunk 3]
```diff
@@ -151,15 +159,20 @@ const EditAttributesModal = ({
...
     component: componentToEdit,
     subcomponents: componentToEdit?.moped_proj_components_subcomponents,
     description: componentToEdit?.description,
     phase: componentToEdit?.moped_phase,
     subphase: componentToEdit?.moped_subphase,
     completionDate: componentToEdit?.completion_date,
+    tags: componentToEdit?.moped_proj_component_tags,
   };
 
+  console.log(initialFormValues)
```
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectComponents/EditAttributesModal.js:174]: ❌

File: moped-database/migrations/1683927220054_run_sql_migration/up.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+ALTER TABLE moped_proj_component_tags ADD CONSTRAINT unique_component_and_tag UNIQUE ("project_component_id", "component_tag_id");
```
Comment (johnclary) [inline moped-database/migrations/1683927220054_run_sql_migration/up.sql:1]: nice!

**Review by johnclary — CHANGES_REQUESTED** [review]
Looks great and works great ✨ 

Just a few minor changes then 🚢 🚢 🚢 🚢 🚢

File: moped-editor/src/views/projects/projectView/ProjectComponents/ComponentForm.js [hunk 5]
```diff
@@ -200,6 +205,16 @@ const ComponentForm = ({
...
+        <Grid item xs={12}>
+          <ControlledAutocomplete
+            id="tags"
+            label="Tags"
+            multiple
+            options={componentTagsOptions}
+            name="tags"
+            control={control}
+          />
+        </Grid>
```
Comment (johnclary) [inline moped-editor/src/views/projects/projectView/ProjectComponents/ComponentForm.js:217]: i think we should up the contrast on the highlights of the already-selected items. I feel like that's out of scope for this issue—i'll make a separate snackoo for it.

Screenshot 2023-05-16 at 2 16 34 PM

File: moped-editor/src/views/projects/projectView/ProjectComponents/utils/form.js (hunk 1 above)
Comment (johnclary) [reply moped-editor/src/views/projects/projectView/ProjectComponents/utils/form.js:115]: as promised: https://github.com/cityofaustin/atd-moped/pull/1051

**Review by roseeichelmann — APPROVED** [review]
test steps worked perfectly ✨

**Review by johnclary — APPROVED** [review]
fantastic—thank you @chiaberry!

you mind holding off on a merge until we bring https://github.com/cityofaustin/atd-moped/pull/1051 along?

edit: ✅ 🚢

**Review by frankhereford — CHANGES_REQUESTED** [review]
Chia, this is great, thank you. It works perfectly when editing, but I think that it's not saving tags from the create component modal. I've got a little recording, if this is helpful. 

https://github.com/cityofaustin/atd-moped/assets/3653206/38ca386a-ed05-4b69-8e0e-9177123c8a7f

Please let me know if I'm missing anything, and I'll flip this to approve in a heartbeat.

Thanks!!

**Comment by chiaberry** [comment]
@frankhereford thank you! i will investigate

**Review by mddilley — COMMENTED** [review]
This all looks great, Chia! Go tags! 🏷️ I noticed that the tag deletes are not persisting after refresh so I'm going to take a quick look at how it works for subcomponents because everything in the code looks in order to me.

File: moped-editor/src/queries/components.js [hunk 6]
```diff
@@ -166,6 +175,7 @@ export const UPDATE_COMPONENT_ATTRIBUTES = gql`
     $phaseId: Int
     $subphaseId: Int
     $completionDate: timestamptz
+    $componentTags: [moped_proj_component_tags_insert_input!]!
   ) {
     update_moped_proj_components_subcomponents(
       where: { project_component_id: { _eq: $projectComponentId } }
```
Comment (mddilley) [inline moped-editor/src/queries/components.js:181]: Just realized what was going on here. This update mutation is first clearing all of the subcomponents by switching any `moped_proj_components_subcomponents` associated with the current component by switch them all to `is_deleted = true`. After that update, we insert the subcomponents chosen in the form, and, if there is a conflict, we update `is_deleted = true` to `is_deleted = false`.  So, the same update for tags is needed here to match up with what happens to subcomponents.

That took me a second to figure out again so let me know if that makes sense. More comments around this with wording from other people than me would make this clearer in the future.

Comment (mddilley) [reply moped-editor/src/queries/components.js:182]: I think that your comment is much clearer! Thanks for updating it. 🙏

**Review by mddilley — APPROVED** [review]
🚢 ✨✨✨

**Review by tillyw — APPROVED** [review]

**Review by frankhereford — APPROVED** [review]
flawlessly

🚀🚀🚀

**Pull request merged**